pyof/foundation/basic_types.py

```diff
@@ -448,22 +442,18 @@ def pack(self, value=None):
         Returns:
             bytes: The binary representation.
 
-        Raises:
-            :exc:`~.exceptions.NotBinaryData`: If value is not :class:`bytes`.
-
         """
-        if isinstance(value, type(self)):
-            return value.pack()
-
         if value is None:
             value = self._value
 
-        if value:
-            if isinstance(value, bytes):
-                return value
-            raise ValueError('BinaryData must contain bytes.')
+        if hasattr(value, 'pack') and callable(value.pack):
+            value = value.pack()
+        elif value == 0:
+            value = b'\x00'
+        elif isinstance(value, list) or value is None:
```

Comment:
You have to check here if `isinstance(value, list)` and also if `len(value)==0`, which means that we have an empty list.

Also, all other cases must have an exception raised, doesn't it? I mean, what if `value` or `_value` is a string?

Please, add a test for it too. BinaryData cannot pack a `string` (but it can pack a `Char` basic type).